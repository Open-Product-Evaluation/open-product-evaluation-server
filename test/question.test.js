const users = require('../seeds/data/user/user')
const surveys = require('../seeds/data/survey/survey')
const questions = require('../seeds/data/question/question')
const { seedDatabase } = require('mongo-seeding')
const config = require('../config')
const request = require('./requesthelper')
const { getSeedID } = require('./helpers')

/* Functions for Querys */

function createQuestionQuery(surveyID, value, description, type) {
  return {
    query: `mutation {
      createQuestion(data:{surveyID:"${surveyID}", value:"${value}", description:"${description}", type:${type}}){
        question{
          type
          items{
            image{
              name
            }
            label
          }
          value
          description
          __typename
        }
      }
    }`,
  }
}

function updateQuestionQuery(questionID, description) {
  return {
    query: `mutation {
      updateQuestion(questionID:"${questionID}", data:{description:"${description}"}){
        question {
          description
        }
      }
    }`,
  }
}

function deleteQuestionQuery(questionID) {
  return {
    query: `mutation {
      deleteQuestion(questionID: "${questionID}") {
        success
      }
    }`,
  }
}

/* Tests */

describe('Question', () => {
  describe('Admin', async () => {
    let jwtToken = ''
    beforeAll(async () => {
      await seedDatabase(config.seeder)
      const expected = users[1]
      const query = {
        query: `mutation {
          login(data: {email: "${expected.email}", password: "password"}) {
            token
            user {
              id
              firstName
              lastName
              email
            }
          }
        }`,
      }
      const { data, errors } = await request.anon(query)
      data.login.token.should.be.a('string')
      expect(errors).toBeUndefined()
      const { login: { token } } = data
      jwtToken = token
    })
    it('should create question [Mutation]', async () => {
      const survey = surveys[1]
      const query = createQuestionQuery(getSeedID(survey), 0, 'TestDescription', 'CHOICE')
      const res = await request.user(query, jwtToken)
      const { data, errors } = res
      expect(data).toMatchSnapshot()
      expect(errors).toBeUndefined()
    })
    it('should update question of survey owned by User [Mutation]', async () => {
      const question = questions[1]
      const query = updateQuestionQuery(getSeedID(question), 'NewTestDescription')
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })
    it('should update question of survey not owned by User [Mutation]', async () => {
      const question = questions[0]
      const query = updateQuestionQuery(getSeedID(question), 'NewTestDescription')
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })
    it('should delete question of survey owned by user [Mutation]', async () => {
      const question = questions[1]
      const query = deleteQuestionQuery(getSeedID(question))
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data.deleteQuestion.success).toBe(true)
    })
    it('should delete question of survey not owned by user [Mutation]', async () => {
      const question = questions[0]
      const query = deleteQuestionQuery(getSeedID(question))
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data.deleteQuestion.success).toBe(true)
    })
  })
  describe('User', async () => {
    let jwtToken = ''
    beforeAll(async () => {
      await seedDatabase(config.seeder)
      const expected = users[0]
      const query = {
        query: `mutation {
          login(data: {email: "${expected.email}", password: "password"}) {
            token
            user {
              id
              firstName
              lastName
              email
            }
          }
        }`,
      }
      const { data, errors } = await request.anon(query)
      data.login.token.should.be.a('string')
      expect(errors).toBeUndefined()
      const { login: { token } } = data
      jwtToken = token
    })
    it('should create question [Mutation]', async () => {
      const query = createQuestionQuery('TestQuestion', 4, 'TestDescription', 'CHOICE')
      const res = await request.user(query, jwtToken)
      const { data, errors } = res
      expect(data).toMatchSnapshot()
      expect(errors).toBeUndefined()
    })
    it('should update question of survey owned by User [Mutation]', async () => {
      const question = questions[0]
      const query = updateQuestionQuery(getSeedID(question), 'NewTestDescription')
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })
    it('should not update question of survey not owned by User [Mutation]', async () => {
      const question = questions[1]
      const query = updateQuestionQuery(getSeedID(question), 'NewTestDescription')
      const { data, errors } = await request.user(query, jwtToken)
      expect(data).toBeNull()
      expect(errors.length).toBeGreaterThan(0)
    })
    it('should delete question of survey owned by user [Mutation]', async () => {
      const question = questions[0]
      const query = deleteQuestionQuery(getSeedID(question))
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data.deleteQuestion.success).toBe(true)
    })
    it('should not delete question of survey not owned by user [Mutation]', async () => {
      const question = questions[1]
      const query = deleteQuestionQuery(getSeedID(question))
      const { data, errors } = await request.user(query, jwtToken)
      expect(data).toBeNull()
      expect(errors.length).toBeGreaterThan(0)
    })
  })
  describe.skip('Device', async () => {
    // TODO Not testable without device login function
    let jwtToken = ''
    beforeAll(async () => {
      await seedDatabase(config.seeder)
      const query = {
        query: `mutation{createDevice(data:{name:"TestDevice"}){
          token
       }}`,
      }
      const { data, errors } = await request.anon(query)
      data.createDevice.token.should.be.a('string')
      expect(errors).toBeUndefined()
      const { createDevice: { token } } = data
      jwtToken = token
    })
    it('should not delete device [Mutation]', async () => {
      const device = questions[3]
      const query = deleteQuestionQuery(getSeedID(device))
      const { data, errors } = await request.user(query, jwtToken)
      expect(data).toBeNull()
      expect(errors.length).toBeGreaterThan(0)
    })
  })
})
