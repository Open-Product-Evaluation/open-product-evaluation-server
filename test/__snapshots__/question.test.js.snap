// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question Admin schould create question [Mutation] 1`] = `
Object {
  "createQuestion": Object {
    "question": Object {
      "__typename": "ChoiceQuestion",
      "description": "TestDescription",
      "items": null,
      "type": "CHOICE",
      "value": "0",
    },
  },
}
`;

exports[`Question Admin schould update question of survey not owned by User [Mutation] 1`] = `
Object {
  "updateQuestion": Object {
    "question": Object {
      "description": "NewTestDescription",
    },
  },
}
`;

exports[`Question Admin schould update question of survey owned by User [Mutation] 1`] = `
Object {
  "updateQuestion": Object {
    "question": Object {
      "description": "NewTestDescription",
    },
  },
}
`;

exports[`Question User schould create question [Mutation] 1`] = `undefined`;

exports[`Question User schould update question of survey owned by User [Mutation] 1`] = `
Object {
  "updateQuestion": Object {
    "question": Object {
      "description": "NewTestDescription",
    },
  },
}
`;
